### PR TITLE
Fix #1993: Schedule a previous cycle reminder on upgrade if needed

### LINKED ISSUE #1993: Show monthly ad grant reminder for previous cycle if upgrade during beginning of new cycle
If the user upgrades & launches at the beginning of a new ad cycle (1st to 7th day of month), we should check to see if they viewed any ads in the previous cycle and schedule an ad reminder on the 7th of the current month for it since they will be getting an ad grant then

### PR DESCRIPTION
**Pull request by kylehickinson**
If the user views ads in the previous cycle upon upgrading, a reminder will be scheduled immediately for the previous cycle

This PR also updates the BraveRewards library, includes this change: https://github.com/brave/brave-core/pull/4030/files

## Summary of Changes

This pull request fixes issue #1993 

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Test Plan:

- Install old build without this PR, view some ads
- Set date between Dec 1 and Dec 6
- Update to new build with this PR
- Launch app, reminder notification should be scheduled for Dec 7, 1:21pm (Can confirm in logs)
- Set date to Dec 7, 1:20, wait 1m and verify a notification appears (in your notification centre! It delivers quietly so you won't see it on lock screen/as a banner)

## Reviewer Checklist:

- [x] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `release-notes/(include|exclude)`
  - `bug` / `enhancement`
- [x] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [x] Adequate unit test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable).
- [x] Issue is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 1]
```diff
@@ -236,6 +236,14 @@ class BrowserViewController: UIViewController {
         
         setupRewardsObservers()
         
+        if let rewards = rewards, !Preferences.Rewards.checkedPreviousCycleForAdsViewing.value {
```
Comment (kylehickinson) [inline Client/Frontend/Browser/BrowserViewController.swift:239]: Ref #1604, the `let rewards = rewards` will have to be removed in that PR

**Review by iccub — APPROVED** [review]

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 1 above)
Comment (iccub) [reply Client/Frontend/Browser/BrowserViewController.swift:239]: I'll rebase and merge the NO_REWARDS PR after this one is merged

File: BraveRewards/BraveRewards.framework/Headers/ledger.mojom.objc.h [hunk 2]
```diff
@@ -158,7 +158,7 @@ typedef NS_ENUM(NSInteger, BATPromotionStatus) {
 
 typedef NS_ENUM(NSInteger, BATPlatform) {
   BATPlatformDesktop = 0,
-  BATPlatformAndroid = 1,
+  BATPlatformAndroidR = 1,
```
Comment (Brandon-T) [inline BraveRewards/BraveRewards.framework/Headers/ledger.mojom.objc.h:161]: Huh? is this intentional? "Android R"?

File: BraveShared/Rewards/MonthlyAdsGrantReminder.swift [hunk 3]
```diff
@@ -42,21 +45,43 @@ final public class MonthlyAdsGrantReminder {
...
+        let currentDay = calendar.component(.day, from: now)
+        switch cycle {
+        case .previous:
+            if currentDay < dayOfReminder {
+                return calendar.component(.month, from: now)
+            }
+            fallthrough
+        case .current:
+            guard let nextMonthsDate = calendar.date(byAdding: .month, value: 1, to: now) else {
+                assertionFailure("Apocalypse...")
```
Comment (Brandon-T) [inline BraveShared/Rewards/MonthlyAdsGrantReminder.swift:62]: lol.

File: BraveRewards/BraveRewards.framework/Headers/ledger.mojom.objc.h (hunk 2 above)
Comment (kylehickinson) [reply BraveRewards/BraveRewards.framework/Headers/ledger.mojom.objc.h:161]: Seems like it 😆

https://github.com/brave/brave-core/blob/00c5caf684aec24668e4e54c68974ea763c5eeb1/vendor/bat-native-ledger/include/bat/ledger/public/interfaces/ledger.mojom#L358

**Pull request merged**